tools/halide_image_io.h

```diff
@@ -310,6 +330,112 @@ bool save_png(ImageType &im, const std::string &filename) {
     png_destroy_write_struct(&png_ptr, &info_ptr);
 
     return true;
+#endif // HALIDE_NOPNG
+}
+
+template<typename ImageType, Internal::CheckFunc check = Internal::CheckReturn>
+bool load_pgm(const std::string &filename, ImageType *im) {
+
+    /* open file and test for it being a pgm */
+    Internal::FileOpener f(filename.c_str(), "rb");
+    if (!check(f.f != nullptr, "File %s could not be opened for reading\n", filename.c_str())) return false;
+
+    int width, height, maxval;
+    char header[256];
+    char buf[1024];
+    bool fmt_binary = false;
+
+    f.readLine(buf, 1024);
+    if (!check(sscanf(buf, "%255s", header) == 1, "Could not read PGM header\n")) return false;
+    if (header == std::string("P5") || header == std::string("p5"))
+        fmt_binary = true;
+    if (!check(fmt_binary, "Input is not binary PGM\n")) return false;
+
+    f.readLine(buf, 1024);
+    if (!check(sscanf(buf, "%d %d\n", &width, &height) == 2, "Could not read PGM width and height\n")) return false;
+    f.readLine(buf, 1024);
+    if (!check(sscanf(buf, "%d", &maxval) == 1, "Could not read PGM max value\n")) return false;
+
+    int bit_depth = 0;
+    if (maxval == 255) { bit_depth = 8; }
+    else if (maxval == 65535) { bit_depth = 16; }
+    else { if (!check(false, "Invalid bit depth in PGM\n")) return false; }
```

Comment:
The braces on this line are in a bizarre place, probably intended:

    else if (!check(false, "Invalid bit depth in PGM\n")) { return false; }